### ISSUE TITLE
Play Asset Delivery Asset Pack ANE

### ISSUE THREAD

**Issue by BMGMobile** [airsdk/Adobe-Runtime-Support#1176]
In section 7.2 "Play Asset Delivery" the release notes for 33.1.1.575 mentioned: `To access the asset packs via the Android Asset Pack Manager functionality, an ANE is being developed.`

Is there any update on this or is there a workaround?

**Comment by marchbold** [comment]
I believe Harman are just making some final changes to this extension and then it should be available shortly. 

What do you mean by a workaround? Are you trying to access asset packs without the ANE?

**Comment by BMGMobile** [comment]
That's good to hear.

By workaround I suppose I was wondering if there is anyone out there who has managed to get this working without the official ANE.  I'd assuming we're forced to call the native AAssetManager api found here: https://developer.android.com/ndk/reference/group/asset#aassetmanager but if there's any secret sauce I'm not aware of I'd love to know :)

**Comment by marchbold** [comment]
Ah okay, as far as I'm aware you need some native calls that the ANE will deliver. I haven't looked into it as Harman are developing one so I left it with them.

**Comment by ajwfrost** [comment]
Yes we're just making some minor changes which should be ready shortly, I'll try to get this uploaded later today..
If you're doing this yourself then yes that's one option, although we're just using the Java API (AssetPackManager):
https://developer.android.com/reference/com/google/android/play/core/assetpacks/AssetPackManager

thanks

**Comment by ajwfrost** [comment]
@BMGMobile  - please extract the below zip to get to our ANE. There's just some minor tweaks needed to document some of the return status codes and then we'll get this formally released with some more documentation.. but for now it should at least help you get started.

[PlayAssetDelivery.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7220946/PlayAssetDelivery.zip)

There are some initial guidance notes available under https://github.com/airsdk/Adobe-Runtime-Support/discussions/1036#discussioncomment-1135665

thanks

**Comment by BMGMobile** [comment]
@ajwfrost Thank you for the ANE.
 I'm having a bit of trouble getting to work using the example code  I've tried installing directly via BundleTool and using the google play internal testing system with the same results.

Running the example code as follows (we only have an install-time asset pack so I've commented out the Fast-Follow/Demand Delivery):
```
var _ane = new PlayAssetDelivery();
var status = _ane.initAssetDelivery(); //this  should be called first and should return status success
trace("\tAAB ANE status:" + status);

// For  loading install time asset 
status =_ane.loadInstallTimeAsset("UI_TooltipAnimations.swf");//this should return status success, here asset-name is the name of the asset file  inside install time asset Pack
trace("\tAAB ANE status:" + status);
/*
// to download or install Fast-follow and Demand Delivery  assetpack  
status =_ane.FetchAsset("AssetPack"); // This method returns  the AssetPackStates initial download states and sizes.
FlashUtil.DevTrace("AAB ANE status:" + status);
*/
// To check the status of assetPack:
var status2 = _ane.getAssetStatus("AssetPack");//this method return the status of assetpack  6- Cancelled,4-Completed,2-Downloading,5-failed,8-Not_installed,1-pending,3-Transferring,0-unknown
trace("\tAAB ANE status:" + status2);

// to get absolute path for an asset:
var path:String = "assets\\UI_TooltipAnimations.swf";
var status3 =_ane.getAssetAbsolutePath("AssetPack", path);//this method returns the absolute path only  if the assetpack state is completed -4 ,otherwise it returns null.path refers to path of the file with asset pack
trace("\tAAB ANE status:" + status3 + ":" + path);
```
 The status return for `_ane.initAssetDelivery();` is always false. Unsurprisingly everything after that also ends up being false/0/null.

I've made the additions to the manifest and afaik all my includes are correct. Do you have any suggestions?

**Comment by ajwfrost** [comment]
Hi @BMGMobile - do you see anything else in the log happening when you call the initAssetDelivery() method? When we run it here, we see:
```
09-30 09:25:14.930  5447  8687 D PackageManager: setEnabledSetting : userId = 0 packageName = air.Main cmp = com.google.android.play.core.assetpacks.AssetPackExtractionService newState = 1 callingPackage = 10523/air.Main
09-30 09:25:14.942  5447  8687 D PackageManager: setEnabledSetting : userId = 0 packageName = air.Main cmp = com.google.android.play.core.assetpacks.ExtractionForegroundService newState = 1 callingPackage = 10523/air.Main
```

Just to confirm, our XML AIR app descriptor file has got:
```
    <extensions>
        <extensionID>com.harman.PlayAssetDelivery</extensionID>
        <extensionID>com.google.android.play</extensionID>
    </extensions>
```
and within the manifest | application section of the Android manifestAdditions:
```
					<service
						android:name="com.google.android.play.core.assetpacks.AssetPackExtractionService"
						android:enabled="false"
						android:exported="true" />
					<service
						android:name="com.google.android.play.core.assetpacks.ExtractionForegroundService"
						android:enabled="false"
						android:exported="false" />
```

Does that match?

thanks

**Comment by BMGMobile** [comment]
@ajwfrost sorry for the delay.  We don't have the `<extensionID>com.google.android.play</extensionID>`  I'm not sure what extension that is.  We can wait on this until the ANE is in its final form; I don't want to block you from making progress because you are busy supporting our pre-testing of the ANE.

**Comment by ajwfrost** [comment]
@BMGMobile that Play ANE is from Distriqt: https://github.com/distriqt/ANE-GooglePlayServices

Hopefully you should be able to see a file `com.google.android.play.ane` within the `lib` folder?

thanks

**Comment by AdamSun** [comment]
@ajwfrost By adding the com.google.android.play.ane and service to the manifest as you mentioned, the status return from _ane.initAssetDelivery() is still false, do I ignore any step?

**Comment by PippoApps** [comment]
@ajwfrost the ANE on the github page seems to be updated on January 2021, this thread though shows an ANE from sept 2021.
Are there any news on this?
I just realised OBB are not suitable anymore and I am stuck.
Thank you

**Comment by ajwfrost** [comment]
Last update was January 21, 2022 as far as I can see?
https://github.com/airsdk/ANE-PlayAssetDelivery/blob/main/build/com.harman.PlayAssetDelivery.ane

**Comment by PippoApps** [comment]
> Last update was January 21, 2022 as far as I can see? https://github.com/airsdk/ANE-PlayAssetDelivery/blob/main/build/com.harman.PlayAssetDelivery.ane
Ah yes!
Thx!
